### PR TITLE
BUG: support ambiguous=infer with ZoneInfo

### LINKED ISSUE #48442: BUG: Series.dt.tz_localize not recognizing DST transition with zoneinfo timezone when ambiguous="infer"
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import pytz
from zoneinfo import ZoneInfo
from datetime import datetime, timedelta

dts = [datetime(2022, 11, 6, 1, 0) + timedelta(minutes = 15*i) for i in range(4)]*2 + [datetime(2022, 11, 6, 2, 0) + timedelta(minutes = 15*i) for i in range(2)]

s = pd.Series(dts)

zoneinfo_infer = s.dt.tz_localize(ZoneInfo('US/Eastern'), ambiguous = 'infer')
print('Result of `tz_localize` with zoneinfo: (all ambiguous times showing UTC -4:00')
print(zoneinfo_infer)
print('-'*80)

pytz_infer = s.dt.tz_localize(pytz.timezone('US/Eastern'), ambiguous = 'infer')
print('Result of `tz_localize` with pytz: (correctly showing DST transition')
print(pytz_infer)


# should raise AmbiguousTimeError, but doesn't
s.dt.tz_localize(ZoneInfo('US/Eastern'))

# does raise AmbiguousTimeError
s.dt.tz_localize(pytz.timezone('US/Eastern'))
```

### Issue Description

On 1.5.0rc0  

Series.dt.tz_localize using a zoneinfo.Zoneinfo timezone does not correctly recognize a daylight savings time transition  

Looks like the cause of this is:
*  `DatetimeArray.tz_localize` calls [tz_localize_to_utc](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/tzconversion.pyx#L187) which instantiates a `Localizer` class as `info`
* `Localizer` sets `self.use_tzlocal = True` [because the tz is zoneinfo](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/tzconversion.pyx#L81)
* `info.use_tzlocal` makes each value [get set using _tz_localize_using_tzinfo_api](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/tzconversion.pyx#L249), bypassing any of the ambiguous time checks
* [_tz_localize_using_tzinfo_api](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/tzconversion.pyx#L598) uses the tzinfo.utcoffset method to [convert the value to UTC (via a round trip to np_datetimestruct to a new naive datetime instance](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/tslibs/tzconversion.pyx#L650)
* The new naive datetime's fold attr is 0 by default, which then always returns Daylight Time during a transition  

### Expected Behavior

Usage of zoneinfo should match that of pytz or dateutil

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 224458ee25d92ccdf289d1ae2741d178df4f323e
python           : 3.10.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19044
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.0rc0
numpy            : 1.23.1
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 63.4.1
pip              : 22.2.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.1
html5lib         : None
pymysql          : 1.0.2
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.4.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.5.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : 2022.1

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #48442 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Sits on top of #49684

### CODE REVIEW

**Comment by mroeschke** [comment]
Fixes https://github.com/pandas-dev/pandas/issues/48442?

File: pandas/_libs/tslibs/tzconversion.pyx [hunk 1]
```diff
@@ -483,6 +489,72 @@ cdef _get_utc_bounds(
...
+        #  need to re-add later as 'extra''
+        extra = (dts.ps // 1000) * (pps // 1_000_000_000)
+
+        dt = datetime_new(dts.year, dts.month, dts.day, dts.hour,
+                          dts.min, dts.sec, dts.us, None)
+
+        aware = dt.replace(tzinfo=tz)
+        as_utc = aware.astimezone(utc_stdlib)
+        rt = as_utc.astimezone(tz)
+        if aware != rt:
```
Comment (mroeschke) [inline pandas/_libs/tslibs/tzconversion.pyx:536]: Might be sufficient to check the `dst` attribute when constructing with `fold=0/1`

```
In [1]: import datetime

In [2]: import zoneinfo

In [3]: before = datetime.datetime(2022, 11, 6, 1, fold=0, tzinfo=zoneinfo.ZoneInfo("US/Pacific"))

In [4]: after = datetime.datetime(2022, 11, 6, 1, fold=1, tzinfo=zoneinfo.ZoneInfo("US/Pacific"))

In [5]: before
Out[5]: datetime.datetime(2022, 11, 6, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='US/Pacific'))

In [6]: after
Out[6]: datetime.datetime(2022, 11, 6, 1, 0, fold=1, tzinfo=zoneinfo.ZoneInfo(key='US/Pacific'))

In [7]: before.dst()
Out[7]: datetime.timedelta(seconds=3600)

In [8]: after.dst()
Out[8]: datetime.timedelta(0)
```

Comment (jbrockmendel) [reply pandas/_libs/tslibs/tzconversion.pyx:536]: IIUC that would check for if the datetime is ambiguous, but i dont think that checks for non-existent

**Comment by jbrockmendel** [comment]
> Fixes https://github.com/pandas-dev/pandas/issues/48442? (and therefore this change is/should be considered for backporting)

yes

File: pandas/_libs/tslibs/tzconversion.pyx (hunk 1 above)
Comment (mroeschke) [reply pandas/_libs/tslibs/tzconversion.pyx:536]: inferring is only supported for the `ambiguous` parameter though?

I think from the ` for i in range(n):` loop below, looks like nonexistent times should just fall though to the `else` clause i.e. nonexistent times shouldn't be populated in `result_a` and `result_b`

Comment (jbrockmendel) [reply pandas/_libs/tslibs/tzconversion.pyx:536]: > looks like nonexistent times should just fall though to the else

For that to happen, `result_a[i]` or `result_b[i]` need to be NPY_NAT; this is where that is getting set.

Comment (jbrockmendel) [reply pandas/_libs/tslibs/tzconversion.pyx:536]: i dont think there is anything actionable here?

**Review by mroeschke — APPROVED** [review]
LGTM. Related side question, do we have testing tz_localize/convert with the  `nonexistent` options?

**Comment by jbrockmendel** [comment]
> do we have testing tz_localize/convert with the nonexistent options?

you mean specific to ZoneInfo?  i dont think so.  would be straightforward to add ZoneInfo case to test_tz_localize_nonexistent since it is currently parametrized over pytz/dateutil

**Comment by mroeschke** [comment]
> > do we have testing tz_localize/convert with the nonexistent options?
> 
> you mean specific to ZoneInfo? i dont think so. would be straightforward to add ZoneInfo case to test_tz_localize_nonexistent since it is currently parametrized over pytz/dateutil

Correct, yes if these can be added in a follow up that would be good.

**Comment by jbrockmendel** [comment]
good call on checking this; ATM this segfaults

**Pull request merged**